### PR TITLE
Allow the scheduler client to contact a scheduler at a non-root URL prefix

### PR DESCRIPTION
**Pull request by stephenpascoe**
We'd like to deploy luigid behind an nginx proxy where the public urls have a prefix.  E.g. `GET /luigi/api/task_list?...`.  This is the minimum change needed to allow a worker to contact luigid in this case.  If you are happy with the feature I can make the required configuration hooks.

### CODE REVIEW

File: luigi/rpc.py [hunk 1]
```diff
@@ -47,9 +47,10 @@ class RemoteScheduler(Scheduler):
     Scheduler proxy object. Talks to a RemoteSchedulerResponder.
     """
 
-    def __init__(self, host='localhost', port=8082, connect_timeout=None):
+    def __init__(self, host='localhost', port=8082, url_prefix='', connect_timeout=None):
```
Comment (gpoulin) [inline luigi/rpc.py]: Can you put the `url_prefix argument` after `connect_timeout` to be backward compatible.

File: luigi/rpc.py [hunk 2]
```diff
@@ -71,7 +72,7 @@ def _async_fetch(self, full_url, body):
 
     def _fetch(self, url, body, log_exceptions=True, attempts=3):
 
-        full_url = 'http://%s:%d%s' % (self._host, self._port, url)
+        full_url = 'http://%s:%d%s%s' % (self._host, self._port, self._url_prefix, url)
```
Comment (gpoulin) [inline luigi/rpc.py]: I think it reach the critical point where we should be using pyfomart.
`{protocol}://{domain}:{port}/{prefix}/{route}`

**Comment by gpoulin** [comment]
Any reason to prefer sub folder over sub domain? It's usually the best way to avoid any issue.

The PR seems OK. However, does that mean that Luigi webserver should be compatible with sub folder. Currently it [isn't](https://github.com/spotify/luigi/blob/master/luigi/templates/header.html). With the current (lack of) security feature, the webserver should be run in a private network where the dns can be controlled. My feeling it's that it add to the complexity but I don't see the benefit. Maybe I'm missing a use case thou.

File: luigi/rpc.py (hunk 2 above)
Comment (Tarrasch) [inline luigi/rpc.py]: While you're at it, you maybe want to change this to use `.format()`. :)

(I thought `%` syntax was deprecated in python 3. No?)

**Comment by Tarrasch** [comment]
Yea, would be nice to include a description of your setup for us to see understand why you did this better. :)

What did you have in mind when you say configuration hooks?

File: luigi/rpc.py (hunk 2 above)
Comment (gpoulin) [inline luigi/rpc.py]: It's what I meant by pyformat (which is in fact something [else](https://www.python.org/dev/peps/pep-0249/#paramstyle)).

(They call it [old %-formatting](https://docs.python.org/3.4/library/string.html#format-examples), but I don't see it going away any time soon.)

Comment (Tarrasch) [inline luigi/rpc.py]: Sorry, I should have read what you wrote more carefully. :)

Yea, the old % formatting is unpythonic and should be avoided.

**Comment by stephenpascoe** [comment]
We prefer a subfolder because we are deploying our own web-app alongside luigi which receives Luigi events and utilises the json api.  So from the users perspective it is one app at one URL.

I agree that it might be better if the Luigi webserver was completely prefix-aware but on brief testing this is enough for us.  We can use nginx URL rewriting to make the webapp work without being aware of a prefix.  This appears to be enough:

    location ~ /luigi(?<path>.*) {
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $http_host;
        proxy_set_header X-NginX-Proxy true;
        proxy_pass http://127.0.0.1:8082$path$is_args$args;
        proxy_redirect off;
        proxy_http_version 1.1;

    }

I have not tested the history feature or the error traceback display yet.

**Comment by aldanor** [comment]
+1, this would be extremely useful

**Comment by erikbern** [comment]
I can merge if you rebase!

**Comment by tashrifbillah** [comment]
Hi @stephenpascoe , to be able to proxy Luigi server against non-root URL `http://pnl_servers.dipr.partners.org/luigi/`. I used your snippet verbatim. Upon plugging in `https://pnl_servers.dipr.partners.org/luigi/` on Mozilla, the URL got rewritten as `http://pnl_servers.dipr.partners.org/static/visualiser/index.html` (no trailing `#` which I expected to see though) and I get an error404. Please see the log below. Can you advise?

<details>

```vim
2020/07/28 10:47:41 [debug] 102110#0: *459 http keepalive handler
2020/07/28 10:47:41 [debug] 102110#0: *459 malloc: 0000561A167EBC60:1024
2020/07/28 10:47:41 [debug] 102110#0: *459 recv: eof:0, avail:1
2020/07/28 10:47:41 [debug] 102110#0: *459 recv: fd:3 391 of 1024
2020/07/28 10:47:41 [debug] 102110#0: *459 reusable connection: 0
2020/07/28 10:47:41 [debug] 102110#0: *459 posix_memalign: 0000561A1680A6F0:4096 @16
2020/07/28 10:47:41 [debug] 102110#0: *459 event timer del: 3: 2682770727
2020/07/28 10:47:41 [debug] 102110#0: *459 http process request line
2020/07/28 10:47:41 [debug] 102110#0: *459 http request line: "GET /static/visualiser/index.html HTTP/1.1"
2020/07/28 10:47:41 [debug] 102110#0: *459 http uri: "/static/visualiser/index.html"
2020/07/28 10:47:41 [debug] 102110#0: *459 http args: ""
2020/07/28 10:47:41 [debug] 102110#0: *459 http exten: "html"
2020/07/28 10:47:41 [debug] 102110#0: *459 posix_memalign: 0000561A16750E60:4096 @16
2020/07/28 10:47:41 [debug] 102110#0: *459 http process request header line
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "Host: pnl_servers.dipr.partners.org"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "Accept-Language: en-US,en;q=0.5"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "Accept-Encoding: gzip, deflate"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "DNT: 1"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "Connection: keep-alive"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "Upgrade-Insecure-Requests: 1"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header: "Cache-Control: max-age=0"
2020/07/28 10:47:41 [debug] 102110#0: *459 http header done
2020/07/28 10:47:41 [debug] 102110#0: *459 generic phase: 0
2020/07/28 10:47:41 [debug] 102110#0: *459 rewrite phase: 1
2020/07/28 10:47:41 [debug] 102110#0: *459 test location: ~ "/luigi(?<path>.*)"
2020/07/28 10:47:41 [debug] 102110#0: *459 using configuration ""
2020/07/28 10:47:41 [debug] 102110#0: *459 http cl:-1 max:1048576
2020/07/28 10:47:41 [debug] 102110#0: *459 rewrite phase: 3
2020/07/28 10:47:41 [debug] 102110#0: *459 post rewrite phase: 4
2020/07/28 10:47:41 [debug] 102110#0: *459 generic phase: 5
2020/07/28 10:47:41 [debug] 102110#0: *459 generic phase: 6
2020/07/28 10:47:41 [debug] 102110#0: *459 generic phase: 7
2020/07/28 10:47:41 [debug] 102110#0: *459 generic phase: 8
2020/07/28 10:47:41 [debug] 102110#0: *459 access phase: 9
2020/07/28 10:47:41 [debug] 102110#0: *459 access phase: 10
2020/07/28 10:47:41 [debug] 102110#0: *459 access phase: 11
2020/07/28 10:47:41 [debug] 102110#0: *459 post access phase: 12
2020/07/28 10:47:41 [debug] 102110#0: *459 generic phase: 13
2020/07/28 10:47:41 [debug] 102110#0: *459 generic phase: 14
2020/07/28 10:47:41 [debug] 102110#0: *459 content phase: 15
2020/07/28 10:47:41 [debug] 102110#0: *459 content phase: 16
2020/07/28 10:47:41 [debug] 102110#0: *459 content phase: 17
2020/07/28 10:47:41 [debug] 102110#0: *459 content phase: 18
2020/07/28 10:47:41 [debug] 102110#0: *459 content phase: 19
2020/07/28 10:47:41 [debug] 102110#0: *459 content phase: 20
2020/07/28 10:47:41 [debug] 102110#0: *459 http filename: "/usr/share/nginx/html/static/visualiser/index.html"
2020/07/28 10:47:41 [debug] 102110#0: *459 add cleanup: 0000561A1680B570
2020/07/28 10:47:41 [error] 102110#0: *459 open() "/usr/share/nginx/html/static/visualiser/index.html" failed (2: No such file or directory), client: 170.223.221.217, server: , request: "GET /static/visualiser/index.html HTTP/1.1", host: "pnl_servers.dipr.partners.org"
2020/07/28 10:47:41 [debug] 102110#0: *459 http finalize request: 404, "/static/visualiser/index.html?" a:1, c:1
2020/07/28 10:47:41 [debug] 102110#0: *459 http special response: 404, "/static/visualiser/index.html?"
2020/07/28 10:47:41 [debug] 102110#0: *459 http set discard body
2020/07/28 10:47:41 [debug] 102110#0: *459 xslt filter header
2020/07/28 10:47:41 [debug] 102110#0: *459 HTTP/1.1 404 Not Found
Server: nginx/1.16.1
Date: Tue, 28 Jul 2020 14:47:41 GMT
Content-Type: text/html
Content-Length: 153
Connection: keep-alive

2020/07/28 10:47:41 [debug] 102110#0: *459 write new buf t:1 f:0 0000561A1680B5F0, pos 0000561A1680B5F0, size: 155 file: 0, size: 0
2020/07/28 10:47:41 [debug] 102110#0: *459 http write filter: l:0 f:0 s:155
2020/07/28 10:47:41 [debug] 102110#0: *459 http output filter "/static/visualiser/index.html?"
2020/07/28 10:47:41 [debug] 102110#0: *459 http copy filter: "/static/visualiser/index.html?"
2020/07/28 10:47:41 [debug] 102110#0: *459 image filter
2020/07/28 10:47:41 [debug] 102110#0: *459 xslt filter body
2020/07/28 10:47:41 [debug] 102110#0: *459 http postpone filter "/static/visualiser/index.html?" 0000561A1680B6B0
2020/07/28 10:47:41 [debug] 102110#0: *459 write old buf t:1 f:0 0000561A1680B5F0, pos 0000561A1680B5F0, size: 155 file: 0, size: 0
2020/07/28 10:47:41 [debug] 102110#0: *459 write new buf t:0 f:0 0000000000000000, pos 0000561A14F964E0, size: 100 file: 0, size: 0
2020/07/28 10:47:41 [debug] 102110#0: *459 write new buf t:0 f:0 0000000000000000, pos 0000561A14F96C40, size: 53 file: 0, size: 0
2020/07/28 10:47:41 [debug] 102110#0: *459 http write filter: l:1 f:0 s:308
2020/07/28 10:47:41 [debug] 102110#0: *459 http write filter limit 0
2020/07/28 10:47:41 [debug] 102110#0: *459 writev: 308 of 308
2020/07/28 10:47:41 [debug] 102110#0: *459 http write filter 0000000000000000
2020/07/28 10:47:41 [debug] 102110#0: *459 http copy filter: 0 "/static/visualiser/index.html?"
2020/07/28 10:47:41 [debug] 102110#0: *459 http finalize request: 0, "/static/visualiser/index.html?" a:1, c:1
2020/07/28 10:47:41 [debug] 102110#0: *459 set http keepalive handler
2020/07/28 10:47:41 [debug] 102110#0: *459 http close request
2020/07/28 10:47:41 [debug] 102110#0: *459 http log handler
2020/07/28 10:47:41 [debug] 102110#0: *459 free: 0000561A1680A6F0, unused: 0
2020/07/28 10:47:41 [debug] 102110#0: *459 free: 0000561A16750E60, unused: 2617
2020/07/28 10:47:41 [debug] 102110#0: *459 free: 0000561A167EBC60
2020/07/28 10:47:41 [debug] 102110#0: *459 hc free: 0000000000000000
2020/07/28 10:47:41 [debug] 102110#0: *459 hc busy: 0000000000000000 0
2020/07/28 10:47:41 [debug] 102110#0: *459 reusable connection: 1
2020/07/28 10:47:41 [debug] 102110#0: *459 event timer add: 3: 65000:2682777104
2020/07/28 10:47:41 [debug] 102110#0: *459 http empty handler
2020/07/28 10:47:41 [debug] 102110#0: timer delta: 6377
2020/07/28 10:47:41 [debug] 102110#0: worker cycle
2020/07/28 10:47:41 [debug] 102110#0: epoll timer: 65000
2020/07/28 10:48:47 [debug] 102110#0: timer delta: 65042
2020/07/28 10:48:47 [debug] 102110#0: *459 event timer del: 3: 2682777104
2020/07/28 10:48:47 [debug] 102110#0: *459 http keepalive handler
2020/07/28 10:48:47 [debug] 102110#0: *459 close http connection: 3
2020/07/28 10:48:47 [debug] 102110#0: *459 reusable connection: 0
2020/07/28 10:48:47 [debug] 102110#0: *459 free: 0000000000000000
2020/07/28 10:48:47 [debug] 102110#0: *459 free: 0000561A1684BE90, unused: 136
2020/07/28 10:48:47 [debug] 102110#0: worker cycle
2020/07/28 10:48:47 [debug] 102110#0: epoll timer: -1

```
</details>

**Pull request merged**